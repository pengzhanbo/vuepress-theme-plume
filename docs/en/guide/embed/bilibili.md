---
title: Bilibili Video
icon: ri:bilibili-fill
createTime: 2025/10/08 12:26:47
permalink: /en/guide/embed/video/bilibili/
---

## Overview

The theme provides the capability to embed Bilibili videos.

This feature is powered by [vuepress-plugin-md-power](../../config/plugins/markdownPower.md).

## Configuration

This feature is disabled by default. You need to enable it in the theme configuration.

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      bilibili: true, // [!code ++]
    },
  })
})
```

## Syntax

Simple syntax:

```md
@[bilibili](bvid)
```

For videos with multiple parts, add `p1`, `p2`, `p3` etc. after `bilibili`:

```md
@[bilibili p1](aid cid)
```

More options:

```md
@[bilibili p1 autoplay time="0" width="100%" height="400px" ratio="16:9"](bvid aid cid)
```

**Parameter Description:**

- bvid: Video BV ID
- aid: Video AID
- cid: Video CID
- autoplay: Whether to enable autoplay
- time: Video playback start time in seconds, or in `mm:ss` or `hh:mm:ss` format
- width: Video width
- height: Video height
- ratio: Video aspect ratio, defaults to `16:9`

For videos with multiple parts, `bvid` can be omitted, but `aid` and `cid` must be provided.

## Examples

### Widescreen Video

Input:

```md
@[bilibili](BV1EZ42187Hg)
```

Output:

@[bilibili](BV1EZ42187Hg)

### Vertical Video

Input:

```md
@[bilibili width="320px" ratio="9:16"](BV1zr42187eg)
```

Output:

@[bilibili width="320px" ratio="9:16"](BV1zr42187eg)
