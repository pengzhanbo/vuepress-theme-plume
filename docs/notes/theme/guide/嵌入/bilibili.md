---
title: Bilibili 视频
icon: ri:bilibili-fill
createTime: 2024/03/28 12:26:47
permalink: /guide/embed/video/bilibili/
---

## 概述

主题提供了 嵌入 Bilibili 视频 的功能。

该功能由 [vuepress-plugin-md-power](../../config/plugins/markdownPower.md) 提供支持。

## 配置

该功能默认不启用。你需要在主题配置中开启。

::: code-tabs
@tab .vuepress/config.ts

```ts
export default defineUserConfig({
  theme: plumeTheme({
    plugins: {
      markdownPower: {
        bilibili: true,
      },
    }
  })
})
```

:::

## 语法

简单的语法：

```md
@[bilibili](bvid)
```

带 分P 的视频，在 `bilibili` 后跟随 `p1`、`p2`、`p3` 等选项

```md
@[bilibili p1](aid cid)
```

更多选项：

```md
@[bilibili p1 autoplay time="0" width="100%" height="400px" ratio="16:9"](bvid aid cid)
```

**选项说明：**

- bvid: 视频 BV ID
- aid: 视频 AID
- cid: 视频 CID
- autoplay: 是否自动播放
- time: 视频开始播放时间点，单位为秒， 也可以使用 `mm:ss` 或 `hh:mm:ss` 格式
- width: 视频宽度
- height: 视频高度
- ratio: 视频比例，默认 `16:9`

对于分P的视频，可以省略传入 `bvid`，但需要传入 `aid` 和 `cid`

## 示例

### 宽频视频

输入：

```md
@[bilibili](BV1EZ42187Hg)
```

输出：

@[bilibili](BV1EZ42187Hg)

### 竖屏视频

输入：

```md
@[bilibili width="320px" ratio="9:16"](BV1zr42187eg)
```

输出：

@[bilibili width="320px" ratio="9:16"](BV1zr42187eg)
