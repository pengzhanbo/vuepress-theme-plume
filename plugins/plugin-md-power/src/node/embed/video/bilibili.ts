/**
 * @[bilibili](bid)
 * @[bilibili](aid cid)
 * @[bilibili](bid aid cid)
 * @[bilibili p1 autoplay time=1](aid cid)
 */
import type { PluginWithOptions } from 'markdown-it'
import type { BilibiliTokenMeta } from '../../../shared/index.js'
import { URLSearchParams } from 'node:url'
import { parseRect } from '../../utils/parseRect.js'
import { resolveAttrs } from '../../utils/resolveAttrs.js'
import { timeToSeconds } from '../../utils/timeToSeconds.js'
import { createEmbedRuleBlock } from '../createEmbedRuleBlock.js'

const BILIBILI_LINK = 'https://player.bilibili.com/player.html'

export const bilibiliPlugin: PluginWithOptions<never> = (md) => {
  createEmbedRuleBlock<BilibiliTokenMeta>(md, {
    type: 'bilibili',
    name: 'video_bilibili',
    // eslint-disable-next-line regexp/no-super-linear-backtracking
    syntaxPattern: /^@\[bilibili(?:\s+p(\d+))?([^\]]*)\]\(([^)]*)\)/,
    meta([, page, info = '', source = '']) {
      const { attrs } = resolveAttrs(info)
      const ids = source.trim().split(/\s+/)
      const bvid = ids.find(id => id.startsWith('BV'))
      const [aid, cid] = ids.filter(id => !id.startsWith('BV'))

      return {
        page: +page || 1,
        bvid,
        aid,
        cid,
        autoplay: attrs.autoplay ?? false,
        time: timeToSeconds(attrs.time),
        title: attrs.title,
        width: attrs.width ? parseRect(attrs.width) : '100%',
        height: attrs.height ? parseRect(attrs.height) : '',
        ratio: attrs.ratio ? parseRect(attrs.ratio) : '',
      }
    },
    content(meta) {
      const params = new URLSearchParams()

      if (meta.bvid) {
        params.set('bvid', meta.bvid)
      }

      if (meta.aid) {
        params.set('aid', meta.aid)
      }

      if (meta.cid) {
        params.set('cid', meta.cid)
      }

      if (meta.page) {
        params.set('p', meta.page.toString())
      }

      if (meta.time) {
        params.set('t', meta.time.toString())
      }

      params.set('autoplay', meta.autoplay ? '1' : '0')

      const source = `${BILIBILI_LINK}?${params.toString()}`

      return `<VideoBilibili src="${source}" width="${meta.width}" height="${meta.height}" ratio="${meta.ratio}" title="${meta.title}" />`
    },
  })
}
