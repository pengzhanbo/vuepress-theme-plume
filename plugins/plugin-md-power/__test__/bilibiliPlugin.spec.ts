import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { bilibiliPlugin } from '../src/node/embed/video/bilibili.js'

function createMarkdown() {
  return MarkdownIt().use((md) => {
    md.block.ruler.before('code', 'import_code', () => false)
    md.renderer.rules.import_code = () => ''
  }).use(bilibiliPlugin)
}

describe('bilibiliPlugin', () => {
  it('should work', () => {
    const md = createMarkdown()
    const code = `\
@[bilibili](BV12345)

@[bilibili](12432 12345)

@[bilibili](BV12345 12343 45678)

@[bilibili p1 autoplay time="1"](BV12345)

@[bilibili p1 autoplay width="100%" height="600px"](BV12345)

@[bilibili p1 autoplay width="100%" ratio="16:9"](BV12345)
`
    expect(md.render(code)).toMatchSnapshot()
  })

  it('should not work', () => {
    const md = createMarkdown()
    const code = `\
@[bilibili]()

@[bilibili]xxx

@[ bilibili]BV12345

@[bilibili](BV12345
`
    expect(md.render(code)).toMatchSnapshot()
  })
})
