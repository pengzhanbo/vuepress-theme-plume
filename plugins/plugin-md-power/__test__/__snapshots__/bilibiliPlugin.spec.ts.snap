// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bilibiliPlugin > should not work 1`] = `
"<VideoBilibili src="https://player.bilibili.com/player.html?autoplay=0" width="100%" height="" ratio="" title="undefined" /><p>@[bilibili]xxx</p>
<p>@[ bilibili]BV12345</p>
<p>@[bilibili](BV12345</p>
"
`;

exports[`bilibiliPlugin > should work 1`] = `"<VideoBilibili src="https://player.bilibili.com/player.html?bvid=BV12345&autoplay=0" width="100%" height="" ratio="" title="undefined" /><VideoBilibili src="https://player.bilibili.com/player.html?aid=12432&cid=12345&autoplay=0" width="100%" height="" ratio="" title="undefined" /><VideoBilibili src="https://player.bilibili.com/player.html?bvid=BV12345&aid=12343&cid=45678&autoplay=0" width="100%" height="" ratio="" title="undefined" /><VideoBilibili src="https://player.bilibili.com/player.html?bvid=BV12345&p=1&t=1&autoplay=1" width="100%" height="" ratio="" title="undefined" /><VideoBilibili src="https://player.bilibili.com/player.html?bvid=BV12345&p=1&autoplay=1" width="100%" height="600px" ratio="" title="undefined" /><VideoBilibili src="https://player.bilibili.com/player.html?bvid=BV12345&p=1&autoplay=1" width="100%" height="" ratio="16:9" title="undefined" />"`;
